src/webServer/webServer.ts

```diff
@@ -0,0 +1,193 @@
+/*@internal*/
+namespace ts.server {
+    export interface HostWithPostMessage {
+        writeMessage(s: any): void;
+    }
+    export interface WebHost extends HostWithPostMessage {
+        readFile(path: string): string | undefined;
+        fileExists(path: string): boolean;
+    }
+
+    export type MessageLogLevel = "info" | "perf" | "error";
+    export interface LoggingMessage {
+        readonly type: "log";
+        readonly level: MessageLogLevel;
+        readonly body: string
+    }
+    export class MainProcessLogger implements Logger {
+        private currentGroupCount = 0;
+        private seq = 0;
+        close = noop;
+
+        constructor(private readonly level: LogLevel, private host: HostWithPostMessage) {
+        }
+
+        hasLevel(level: LogLevel): boolean {
+            return this.level >= level;
+        }
+
+        loggingEnabled(): boolean {
+            return true;
+        }
+
+        perftrc(s: string): void {
+            this.msg(s, Msg.Perf);
+        }
+
+        info(s: string): void {
+            this.msg(s, Msg.Info);
+        }
+
+        err(s: string) {
+            this.msg(s, Msg.Err);
+        }
+
+        startGroup(): void {
+            ++this.currentGroupCount;
+        }
+
+        endGroup(): void {
+            this.currentGroupCount = Math.max(0, this.currentGroupCount - 1);
+        }
+
+        msg(s: string, type: Msg = Msg.Err): void {
+            s = `${type} ${this.seq.toString()} [${nowString()}] ${s}`;

```

Comment:
I assume this is retained from the existing code, but it seems strange to me that the logger wouldn't check the level here, rather than having every call-site do it.

Reply:
This is how logger was designed way back so any change here would be breaking. (esp because msg is suppose to be public method)